modules/adotBidAdapter.js

```diff
@@ -0,0 +1,636 @@
+  if (bid.adm) {
+    if (!isString(bid.adm)) return false;
+    if (!bid.adm.includes(AUCTION_PRICE)) return false;
+  }
+  if (bid.nurl) {
+    if (!isString(bid.nurl)) return false;
+    if (!bid.nurl.includes(AUCTION_PRICE)) return false;
+  }
+
+  if (isBidABanner(bid)) {
+    if (!isNumber(bid.h)) return false;
+    if (!isNumber(bid.w)) return false;
+  }
+  if (isBidAVideo(bid)) {
+    if (!(isNone(bid.h) || isNumber(bid.h))) return false;
+    if (!(isNone(bid.w) || isNumber(bid.w))) return false;
+  }
+
+  const impression = getImpressionData(serverRequest, bid.impid);
+
+  if (!isObject(impression.openRTB)) return false;
+  if (!isObject(impression.internal)) return false;
+  if (!isString(impression.internal.adUnitCode)) return false;
+
+  if (isBidABanner(bid)) {
+    if (!isObject(impression.openRTB.banner)) return false;
+  }
+  if (isBidAVideo(bid)) {
+    if (!isObject(impression.openRTB.video)) return false;
+  }
+  if (isBidANative(bid)) {
+    if (!isObject(impression.openRTB.native) || !tryParse(bid.adm)) return false;
+  }
+
+  return true;
+}
+
+function isBidABanner(bid) {
+  return isObject(bid) &&
+    isObject(bid.ext) &&
+    isObject(bid.ext.adot) &&
+    bid.ext.adot.media_type === 'banner';
+}
+
+function isBidAVideo(bid) {
+  return isObject(bid) &&
+    isObject(bid.ext) &&
+    isObject(bid.ext.adot) &&
+    bid.ext.adot.media_type === 'video';
+}
+
+function isBidANative(bid) {
+  return isObject(bid) &&
+    isObject(bid.ext) &&
+    isObject(bid.ext.adot) &&
+    bid.ext.adot.media_type === 'native';
+}
+
+function getImpressionData(serverRequest, impressionId) {
+  const openRTBImpression = serverRequest.data.imp.find(imp => imp.id === impressionId);
+  const internalImpression = serverRequest._adot_internal.impressions.find(imp => imp.impressionId === impressionId);
+
+  return {
+    id: impressionId,
+    openRTB: openRTBImpression || null,
+    internal: internalImpression || null
+  };
+}
+
+function generateAdFromBid(bid, bidResponse, serverRequest) {
+  const impressionData = getImpressionData(serverRequest, bid.impid);
+  const base = {
+    requestId: impressionData.internal.bidId,
+    cpm: bid.price,
+    currency: bidResponse.cur,
+    ttl: TTL,
+    creativeId: bid.crid,
+    netRevenue: NET_REVENUE,
+    mediaType: bid.ext.adot.media_type,
+  };
+
+  if (isBidANative(bid)) return {...base, native: formatNativeData(bid.adm)};
+
+  const size = getSizeFromBid(bid, impressionData);
+  const creative = getCreativeFromBid(bid, impressionData);
+
+  return {
+    ...base,
+    height: size.height,
+    width: size.width,
+    ad: creative.markup,
+    adUrl: creative.markupUrl,
+    vastXml: null,
+    vastUrl: null,
+    renderer: creative.renderer
+  };
+}
+
+function getVastXml(ad, done) {
+  if (isString(ad.adUrl)) {
+    window.fetch(ad.adUrl)
+      .then(response => response.text())
+      .then(vastXml => done(null, vastXml))
+      .catch(err => done(err, null));
+
+    return;
```

Comment:
I'm afraid we don't support Promises in Prebid.js.  Can you please rewrite this logic without promises?  

Additionally, (if I'm understanding the larger workflow) would it be possible to do this overall logic for your outstream player without it requiring an extra network request just to download the VAST XML?  If there's some way the raw VAST could be returned as part of the bid response from your endpoint, that would be ideal.